File: src/features/quote/services/zoneNotFoundError.ts
/**
 * Raised when the destination country has no zone in the selected carrier's
 * official zone table. There is deliberately no fallback zone — an unmapped
 * destination must surface to the user, not be priced off a guessed zone.
 * Mirrored by Calculators::ZoneNotFoundError (smart-quote-api).
 */
export class ZoneNotFoundError extends Error {
  readonly carrier: 'UPS' | 'DHL' | 'FEDEX';
  readonly country: string;

  constructor(carrier: 'UPS' | 'DHL' | 'FEDEX', country: string) {
    super(`No ${carrier} zone mapping for destination country "${country}"`);
    this.name = 'ZoneNotFoundError';
    this.carrier = carrier;
    this.country = country;
  }
}

File: smart-quote-api/app/services/calculators/zone_not_found_error.rb
# frozen_string_literal: true

module Calculators
  # Raised when the destination country has no zone in the selected carrier's
  # official zone table. There is deliberately no fallback zone — an unmapped
  # destination must become a 422 (ZONE_NOT_FOUND) on the API, never a quote
  # priced off a guessed zone. Mirrors the frontend ZoneNotFoundError.
  class ZoneNotFoundError < StandardError
    attr_reader :carrier, :country

    def initialize(carrier:, country:)
      @carrier = carrier
      @country = country
      super(%(No #{carrier} zone mapping for destination country "#{country}"))
    end
  end
end
